File: apps/bot-python/keyword_match.py
"""Shared keyword matching for taxonomy rules (mirrors App\\Support\\KeywordMatch).

Short tokens (≤3 chars, no space) use letter-boundaries so les ≠ sales and
idi ≠ pendidikan. Longer tokens and phrases stay substring matches (ngopi still
matches kopi).
"""

from __future__ import annotations

import re


def keyword_in(haystack: str, needle: str) -> bool:
    haystack = haystack.lower()
    token = needle.lower().strip()
    if not token:
        return False
    if " " in token or len(token) > 3:
        return token in haystack
    return bool(re.search(rf"(?<![a-z_]){re.escape(token)}(?![a-z_])", haystack))


def any_keyword(haystack: str, needles: tuple[str, ...] | list[str]) -> bool:
    return any(keyword_in(haystack, n) for n in needles)

File: apps/admin-laravel/app/Support/KeywordMatch.php
<?php

namespace App\Support;

/**
 * Phrase matching for taxonomy rules.
 * Short tokens (≤3 chars, no space) use letter-boundaries so "les" ≠ "sales"
 * and "idi" ≠ "pendidikan". Longer / multi-word phrases stay substring matches.
 */
final class KeywordMatch
{
    public static function contains(string $haystack, string $needle): bool
    {
        $haystack = mb_strtolower($haystack);
        $needle = mb_strtolower(trim($needle));
        if ($needle === '') {
            return false;
        }

        if (str_contains($needle, ' ') || mb_strlen($needle) > 3) {
            return str_contains($haystack, $needle);
        }

        // Batas huruf saja: "pph21" masih cocok "pph"; "sales" tidak cocok "les".
        return preg_match(
            '/(?<![\p{L}_])'.preg_quote($needle, '/').'(?![\p{L}_])/u',
            $haystack,
        ) === 1;
    }

    /**
     * @param  list<string>  $needles
     */
    public static function containsAny(string $haystack, array $needles): bool
    {
        foreach ($needles as $needle) {
            if (self::contains($haystack, (string) $needle)) {
                return true;
            }
        }

        return false;
    }
}
